### PR TITLE
refactor(sql): split type checking into CoreExpr modules

### PR DESCRIPTION
**Pull request by forsaken628**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Refactor SQL type checking to lower AST expressions into CoreExpr first, then resolve them through module-specific type-check implementations. This splits the old monolithic
  type_check.rs into focused modules for aggregate, scalar, special functions, UDF, column, variant, window, in-list, SRF, and related behavior.

  Also narrows runtime/catalog/UDF/session dependencies behind the type-check adapter boundary, while preserving existing behavior for edge cases such as IN-list rewrite
  thresholds, SRF restrictions, aggregate/window legality, UDF resolution, variant virtual columns, and vector function fallback.

  Tests were reorganized to match the implementation module split and expanded with golden behavior coverage for the refactored paths.

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/19829)

### CODE REVIEW

File: src/query/sql/src/planner/semantic/type_check/date.rs [hunk 1]
```diff
@@ -26,12 +27,321 @@ use databend_common_expression::Expr as EExpr;
...
+                (Between, Quarter) => "between_quarters",
+                (Between, Month) => "between_months",
+                (Between, Day) => "between_days",
+                (Between, Hour) => "between_hours",
+                (Between, Minute) => "between_minutes",
+                (Between, Second) => "between_seconds",
+                (Between, Doy) => "between_doys",
+                (Between, Week) => "between_weeks",
+                (Between, Dow) => "between_dows",
+                (Between, Epoch) => "between_dows",
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check/date.rs]: should be  "between_seconds" ?

File: src/query/sql/src/planner/semantic/type_check/adapter.rs [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+            .cloud_control_grpc_server_address
+            .is_some()
+        {
+            Some(CloudControlApiProvider::instance())
+        } else {
+            None
+        };
+
+        Self {
+            async_runtime_handle: tokio::runtime::Handle::current(),
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check/adapter.rs]: we should not initialize the async_runtime_handle here, `current` could panic if current thread is not in tokio runtime context.

We should use try_current and call async resolve lazily.

```
fn block_on<F>(&self, future: F) -> Result<F::Output>
    where
        F: std::future::Future,
    {
        let handle = tokio::runtime::Handle::try_current().map_err(|err| {
            ErrorCode::Internal(format!(
                "type check operation requires a tokio runtime: {err}"
            ))
        })?;

        Ok(databend_common_base::runtime::block_on_with_handle(&handle, future))
    }
```

**Review by sundy-li — CHANGES_REQUESTED** [review]

**Review by sundy-li — APPROVED** [review]
